Added functions to interact with JuMP

caioluke opened the pull request:
closes #1 

- Creation of necessary functions in order to interact with the JuMP interface, e.g MOI.is_empty, MOI.get, MOI.set, MOI.supports

- Creation of mutable struct 'ParameterValue' which is used to assign value to parameters

- Correction of a bug in which a ScalarQuadraticFunction was not treated as a ScalarAffineFunction when the quadratic term was a product of 2 parameters

- Added some tests with JuMP

```julia
optimizer = POI.ParametricOptimizer(GLPK.Optimizer())
    

c = [4.0, 3.0]
A1 = [2.0, 1.0, 1.0]
A2 = [1.0, 2.0, 1.0]
b1 = 4.0
b2 = 4.0

# ===== Create model with parameters =====
model = direct_model(optimizer)

@variable(model, x[i=1:2] >= 0)

# Parameters set to 0
@variable(model, y in POI.Parameter(0))
@variable(model, w in POI.Parameter(0))
@variable(model, z in POI.Parameter(0))

@constraint(model, 2*x[1] + x[2] + y <= 4)
@constraint(model, x[1] + 2*x[2] + z <= 4)

@objective(model, Max, sum(c[i]*x[i] for i=1:2) + w)

optimize!(model)
println(model)

Max 4 x[1] + 3 x[2]
Subject to
 x[1] >= 0.0
 x[2] >= 0.0
 x[2] + 2 x[1] <= 4.0
 2 x[2] + x[1] <= 4.0

# ===== Set parameter y to 2 =====
# A better way of assigning values to parameters
MOI.set(model, POI.ParameterValue(), y, 2.0)

optimize!(model)
println(model)

Max 4 x[1] + 3 x[2]
Subject to
 x[1] >= 0.0
 x[2] >= 0.0
 x[2] + 2 x[1] <= 2.0
 3 x[1] + 2 x[2] <= 4.0
```

joaquimg commented:
@caioluke 

I have an ideia to make it better.
We can make a:
`struct ParameterValue <: MOI.AbstractVariableAttribute end`
Then we can simply call:
```julia
MOI.set(model, POI.ParameterValue(), y, 2.0)
```
instead of:
```julia
cy = MOI.get(backend(model), POI.ParameterRef(), index(y))
MOI.set(backend(model), MOI.ConstraintSet(), cy, POI.Parameter(2.0))
```

@odow @blegat 

some questions for you too...

should we add JuMP as a Test/Doc dependency and avoid as much as we can add it as an actual dependency?

Other suggestions?

odow commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,105 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+
+function MOI.empty!(model::ParametricOptimizer{T}) where T
+    MOI.empty!(model.optimizer)
+    model.parameters = Dict{MOI.VariableIndex, Float64}()
```
```suggestion
    empty!(model.parameters)
```

You should be able to do this for the rest of the `Dicts.

caioluke replied:
Well noticed, I will change it

odow commented:
> struct ParameterValue <: MOI.AbstractVariableAttribute end
> Then we can simply call:
>
> MOI.set(model, POI.ParameterValue(), y, 2.0)

:+1: Yeah this would be better.

caioluke commented:
@joaquimg @odow

Would you rather see the parameters in the model, like so

'''
julia

Max 4 x[1] + 3 x[2] + w
Subject to
 2 x[1] + x[2] + y <= 4.0
 x[1] + 2 x[2] + z <= 4.0
 x[1] >= 0.0
 x[2] >= 0.0
 y in Main.ParametricOptInterface.Parameter(0.0)
 w in Main.ParametricOptInterface.Parameter(0.0)
 z in Main.ParametricOptInterface.Parameter(0.0)
''' 

Or would you rather not see the parameters at all as they would assume their values, like so

'''
julia

Max 4 x[1] + 3 x[2]
Subject to
 x[1] >= 0.0
 x[2] >= 0.0
 x[2] + 2 x[1] <= 4.0
 2 x[2] + x[1] <= 4.0
''' 

?

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+
+function MOI.supports_constraint(
+    model::ParametricOptimizer,
+    F::Type{<:MOI.AbstractFunction},
```
Replace `MOI.AbstractFunction` with `Union{MOI.SingleVariable, ...}`. Here, you say you support custom functions defined by MOI extensions but it's not true.

blegat replied:
You also don't support `VectorQuadraticFunction`

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+            MOI.ObjectiveFunction{MOI.ScalarAffineFunction{T}},
+            }) where T
+    return MOI.supports(model.optimizer, attr)
+end
```
Where is `supports` for quadratic function ?

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+    MOI.get(model.optimizer, attr, c)
+end
+
+function MOI.supports(model::ParametricOptimizer, attr::MOI.ConstraintName, tp::Type{MOI.ConstraintIndex})
```
`Type{MOI.ConstraintIndex}` -> `Type{<:MOI.ConstraintIndex}`

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+end
+
+function MOI.get(model::ParametricOptimizer, attr::MOI.ConstraintSet, ci::MOI.ConstraintIndex{F, S}) where {F, S}  
+    if MOI.is_valid(model.optimizer, ci)
```
Use `MOI.throw_if_not_value(model, ci)`

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+end 
+
+function MOI.get(model::ParametricOptimizer, attr::MOI.ConstraintFunction, ci::MOI.ConstraintIndex{F, S}) where {F, S}
+    MOI.get(model.optimizer, attr, ci)
```
This is not correct, you need to put the parameters back into the function. You can leave this for a later PR if you want by leaving a TODO or removing this method or commenting it out with a TODO

jd-lara replied:
@blegat, can you clarify this point? I am working on #59 that needs this method to pass

odow replied:
The function returned by `model.optimizer` will contain the `Float64` value instead of the `VariableIndex` value of the parameter.

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+end
+
+function MOI.get(model::ParametricOptimizer, attr::MOI.ObjectiveFunctionType)
+    MOI.get(model.optimizer, attr)
```
This is not correct, you might  have transformed a quadratic function into an affine function, you need to give the type that was given by the user, not the type of the inner model.

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+    model::ParametricOptimizer,
+    attr::MOI.ObjectiveFunction{F}) where F <: Union{MOI.SingleVariable,MOI.ScalarAffineFunction{T}} where T
+
+    MOI.get(model.optimizer, attr)
```
Same as ConstraintFunction getter. And you also need to convert to `F`

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+function MOI.get(model::ParametricOptimizer, ::MOI.ListOfConstraints)
+    constraints = Set{Tuple{DataType, DataType}}()
+    inner_ctrs = MOI.get(model.optimizer, MOI.ListOfConstraints())
+    for (F, S) in inner_ctrs
```
You might have transformed quadratic functions into affine functions so this is incorrect

blegat commented on src/ParametricOptInterface.jl:
```diff
@@ -70,6 +72,191 @@ mutable struct ParametricOptimizer{T, OT <: MOI.ModelLike} <: MOI.AbstractOptimi
+    MOI.ScalarAffineFunction{T},
+    MOI.VectorAffineFunction{T},
+}} where T
+    MOI.get(model.optimizer, MOI.ListOfConstraintIndices{F, S}())
```
Same as above, some of these might have been quadratic functions

blegat reviewed (commented):
Looking good overall, thanks! There are however a few that are currently incorrect, I would recommend at least leaving at TODO explaining why it's incorrect

caioluke commented:
@blegat with @guilhermebodin we addressed most of your comments and we left some TODO's for the future. If you have any tips on how to implement those changes that would be great.
